Typings break with TypeScript 2.9 keyof change

<!-- BUG TEMPLATE -->
## Environment

* Archlinux, with node 8 via nvm
* styled-components 3.3.0
* typescript 2.9.1

## Reproduction

## Steps to reproduce

Just run `typescript@2.9` in any project that has `styled-components`
```sh
> tsc
```

## Expected Behavior

things should just work

## Actual Behavior

```
node_modules/styled-components/typings/styled-components.d.ts:76:48 - error TS2344: Type 'keyof T' does not satisfy the constraint 'string'.                                                                                                   
  Type 'string | number | symbol' is not assignable to type 'string'.                                                                                                                                                                          
    Type 'number' is not assignable to type 'string'.                                                                                                                                                                                          
                                                                                                                                                                                                                                               
76 type Omit<T, K extends keyof T> = Pick<T, Diff<keyof T, K>>;                                                                                                                                                                                
                                                  ~~~~~~~                                                                                                                                                                                      
```

## Fixes

There are two PRs that fix this problem:
#1684 
#1771 

## The dilemma:

* if you update the type definition, it’s a breaking change because you force TS >= 2.8 on users.
* if you *don’t* update the type definition, things break with TS 2.9, so you are implicitly forcing TS < 2.9 on your users.

--- a/typings/styled-components.d.ts
+++ b/typings/styled-components.d.ts
@@ -72,7 +72,8 @@ export interface ThemedCssFunction<T> {
 }
 
 // Helper type operators
-type Diff<T extends string, U extends string> = ({ [P in T]: P } & { [P in U]: never } & { [x: string]: never })[T];
+type KeyofBase = keyof any;
+type Diff<T extends KeyofBase, U extends KeyofBase> = ({ [P in T]: P } & { [P in U]: never } & { [x: string]: never })[T];
 type Omit<T, K extends keyof T> = Pick<T, Diff<keyof T, K>>;
 type WithOptionalTheme<P extends { theme?: T; }, T> = Omit<P, "theme"> & { theme?: T; };
 

